File: kotlin-node/karakum/src/jsMain/resources/plugins/convertAssertStrictNames.js
import ts from "typescript"

const renamedFunctions = {
    strictEqual: "equal",
    notStrictEqual: "notEqual",
    deepStrictEqual: "deepEqual",
    notDeepStrictEqual: "notDeepEqual",
}

export default function (node, context, render) {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

    if (
        sourceFileName.endsWith("assert.d.ts")
        && ts.isIdentifier(node)
        && node.text in renamedFunctions

        && node.parent
        && ts.isFunctionDeclaration(node.parent)
        && node.parent.name === node
    ) {
        return renamedFunctions[node.text]
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertAssertStrictNames.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.isFunctionDeclaration
import typescript.isIdentifier

private val renamedFunctions = mapOf(
    "strictEqual" to "equal",
    "notStrictEqual" to "notEqual",
    "deepStrictEqual" to "deepEqual",
    "notDeepStrictEqual" to "notDeepEqual",
)

val convertAssertStrictNames = createPlugin { node, _, _ ->
    nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("assert.d.ts"))

        ensure(isIdentifier(node))
        ensure(node.text in renamedFunctions)

        val function = node.parent
        ensure(isFunctionDeclaration(function))
        ensure(function.name == node)

        renamedFunctions[node.text]
    }
}
